### PR TITLE
feat(client): pagination on timeline

### LINKED ISSUE #37412: Profile page - Timeline - Pagination
**Is your feature request related to a problem? Please describe.**
Currently we display almost all the completed challenges in the timeline which means for some users they have to scroll a lot to reach the footer. 

**Describe the solution you'd like**
1) implementing pagination on the time line would make the profile page look more compact. Displaying max of 10 or 20 per page would be nice.
2) I had wanted to suggest to put a pagination on the API side itself but it seems we are fetching all the user data once and using it in other routes. So it seems it's not possible. If it's possible to do, please notify.

Please give your valuable insights and If it's something to be implemented. I would love to help.

### PR DESCRIPTION
**Pull request by MathewtheCoder**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #37412

### CODE REVIEW

**Comment by MathewtheCoder** [comment]
Gif showing the working of pagination in timeline.
Timeline-pagination

File: client/src/components/profile/components/TimeLine.js [hunk 1]
```diff
@@ -81,9 +98,30 @@ class TimelineInner extends Component {
     }));
   }
 
+  nextPage() {
+    this.setState(state => ({
+      ...state,
```
Comment (RandellDawson) [inline client/src/components/profile/components/TimeLine.js]: Why is this line needed?

File: client/src/components/profile/components/TimeLine.js [hunk 2]
```diff
@@ -81,9 +98,30 @@ class TimelineInner extends Component {
...
+      pageNo: state.pageNo + 1
+    }));
+  }
+
+  prevPage() {
+    this.setState(state => ({
+      ...state,
```
Comment (RandellDawson) [inline client/src/components/profile/components/TimeLine.js]: Why is this line needed?

**Review by RandellDawson — CHANGES_REQUESTED** [review]
See my questions.

File: client/src/components/profile/components/TimeLine.js (hunk 1 above)
Comment (MathewtheCoder) [reply client/src/components/profile/components/TimeLine.js]: @RandellDawson If you don't mind can you please explain what the issue is, I didn't quite understand your question thats why ?

Comment (RandellDawson) [reply client/src/components/profile/components/TimeLine.js]: The whole point of setState, is you can specify which properties you want to change.  There is no reason to add the `...state,` because the other properties in state will not be effected anyway.

**Comment by MathewtheCoder** [comment]
@RandellDawson Made the changes you requested. Thanks for pointing it out.

**Comment by ojeytonwilliams** [comment]
Hey @MathewtheCoder, I noticed that the PropTypes were not quite right and the pagination was missing a page.  It seemed easier to make those changes than ask you to. I hope that's okay!  Let me know if I did anything silly.

One other thing I noticed, but didn't touch, was that the pagination buttons are not centred unless both `next` and `prev` are visible.  It might look better if the numbers were centred and the buttons did not move.  @RandellDawson @moT01 which do you prefer?

Finally, when that's decided, it would be great to have some tests.  Something to confirm that it has the correct buttons at the correct times, that sort of thing.

**Comment by MathewtheCoder** [comment]
Thanks @ojeytonwilliams for the help. I will look into the issue you pointed out.

**Comment by MathewtheCoder** [comment]
@ojeytonwilliams @raisedadead 
Below are the changes done in the latest commit.
1) I have implemented the change to make sure the buttons stay centered in all scenarios.
2)I have applied a fixed height to the timeline rows to prevent sudden jerk while moving between pages.
3) I have vertically aligned the contents in timeline tables cell.
Please share your thoughts on this.
Nov-02-2019 15-19-57

**Review by moT01 — COMMENTED** [review]
Overall, this looks great @MathewtheCoder 🎉 

I have a few suggestions that may or may not be better or worth changing...

1. The buttons feel squished up next to the numbers, I would try and make the space between the buttons and numbers the same as the space between the numbers and the word `of`...
Screen Shot 2019-11-02 at 7 31 23 AM

2. Do we need the word and the `<`/`>` in the buttons? Just throwing this out there - it looks good the way it is, but it might look cleaner without the symbols - I'm not sure.

3. I have 54 pages, which means I have to make 53 clicks to get to the last page; and if I were to finish all the challenges, it might be double as many pages. Wondering if we should add a `first` and `last` button - and maybe even added based on a condition, meaning something like... `if pages>10` => `display first and last buttons` for example.

**Comment by MathewtheCoder** [comment]
@moT01 Great suggestions. Here are my views.

1. Yes, it is a bit squished together. I will apply the better spacing.
2. Can i go with <code><</code>/<code>></code> instead of names. Thats would save the translation efforts also.
3. I also felt it was inconvenient navigating through the pages when there's a lot. There are two methods to solve this.
- Conditionally show the first and last button when there is lot of buttons.
- Give a number input to let users input which page they want to go to.
image

**What are your thoughts ?**

**Comment by RandellDawson** [comment]
I know you have been shooting for pagination, but what about lazy loading?  As the user scrolls down towards the bottom of the page, more information is shown.  Users are very familiar with this type of loading, so there would not be a learning curve.

**Comment by moT01** [comment]
If we're going to keep all the data on one page, I think it's good the way it is. If we do some sort of lazy loading; my experience to get to the bottom would be.... scroll a little, load the next section, and repeat a whole bunch. Right now, I can just scroll to the bottom with a click and drag of the scroll bar. I think lazy loading has some "cool factor" going for it, but the data is all already loaded... my opinion anyway --- although we do have that on the `news` area; I see the same problem over there, If I want to get to the bottom, or even three months back, it's a real pain.

<hr>

> Can i go with </> instead of names. That would save the translation efforts also.

I would stick with text, [see this comment](https://github.com/freeCodeCamp/freeCodeCamp/pull/35485#pullrequestreview-248465838)

>I also felt it was inconvenient navigating through the pages when there's a lot. There are two methods to solve this.

> Conditionally show the first and last button when there is lot of buttons as you suggested.

> Give a number input to let users input which page they want to go to.

I'm not sure on this, an input might not be obvious enough

**Comment by MathewtheCoder** [comment]
@moT01 
1) For the button i am using [aria-label](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute). Currently its placed for li element which is wrong and it should be given to the buttons. So I believe this will make sure the *a11y* is not affected like @raisedadead said in the above comment you shared. So we can use the <code><</code>/<code>></code> without worrying about a11y.

> Conditionally show the first and last button when there is lot of buttons as you suggested.

I will go with this approach then.

**Comment by MathewtheCoder** [comment]
@moT01 I have made the changes.

File: client/src/components/profile/components/TimeLine.js [hunk 3]
```diff
@@ -39,20 +60,26 @@ class TimelineInner extends Component {
 
     this.state = {
       solutionToView: null,
-      solutionOpen: false
+      solutionOpen: false,
+      pageNo: 1,
+      totalPages: Math.ceil(props.sortedTimeline.length / ITEMS_PER_PAGE)
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimeLine.js]: This is the only thing I'm not 100% on.  I think it would be better to use state to hold, well, just the current state of the page.  `totalPages` does not change unless the props change, so it would be a bit better to pass in `totalPages` to `TimelineInner` as a separate prop.  That way there's still a single source of truth and state is reserved for recording user actions.

Possibly I'm being overly restrictive about how state can be used, though.

I'm probably being overly pedantic, though, so

File: client/src/components/profile/components/TimelinePagination.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      <ul className='timeline-pagination_list'>
+        {totalPages > 10 && (
+          <li
+            className='timeline-pagination_list_item'
+            style={{
+              visibility: pageNo === 1 ? 'hidden' : 'unset'
+            }}
+          >
+            <button
+              aria-label='Goto First page'
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimelinePagination.js]:
```suggestion
              aria-label='Go to first page'
```
Goto is just a computing term, as far as I know.  Presumably it would be pronounced differently from 'go to' by screen readers.

File: client/src/components/profile/components/TimelinePagination.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+          </li>
+        )}
+        <li
+          className='timeline-pagination_list_item'
+          style={{
+            visibility: pageNo === 1 ? 'hidden' : 'unset'
+          }}
+        >
+          <button
+            aria-label='Goto Previous page'
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimelinePagination.js]:
```suggestion
            aria-label='Go to previous page'
```

File: client/src/components/profile/components/TimelinePagination.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+          {pageNo} of {totalPages}
+        </li>
+        <li
+          className='timeline-pagination_list_item'
+          style={{
+            visibility: pageNo === totalPages ? 'hidden' : 'unset'
+          }}
+        >
+          <button
+            aria-label='Goto Next page'
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimelinePagination.js]:
```suggestion
            aria-label='Go to next page'
```

File: client/src/components/profile/components/TimelinePagination.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        </li>
+        {totalPages > 10 && (
+          <li
+            className='timeline-pagination_list_item'
+            style={{
+              visibility: pageNo === totalPages ? 'hidden' : 'unset'
+            }}
+          >
+            <button
+              aria-label='Goto Last page'
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimelinePagination.js]:
```suggestion
              aria-label='Go to last page'
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
It's looking really nice, all in all.  I pointed out a slight change to the architecture that you might want to consider.

Also there's some nitpicking about 'go to'.

File: client/src/components/profile/components/TimeLine.js (hunk 3 above)
Comment (MathewtheCoder) [reply client/src/components/profile/components/TimeLine.js]: Thanks for the suggestion. I have made totalPages as a prop instead of state.

**Comment by MathewtheCoder** [comment]
@ojeytonwilliams @RandellDawson I have made the changes requested.

File: client/src/components/profile/components/TimeLine.js [hunk 8]
```diff
@@ -81,9 +108,39 @@ class TimelineInner extends Component {
...
+    }));
+  }
+
+  prevPage() {
+    this.setState(state => ({
+      pageNo: state.pageNo - 1
+    }));
+  }
+  lastPage() {
+    this.setState({
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimeLine.js]:
```suggestion
    this.setState((_, props) => ({
      pageNo: props.totalPages
    }));
```
Just to be extra careful, since `setState` calls can be batched.

File: client/src/components/profile/components/TimeLine.js [hunk 9]
```diff
@@ -81,9 +108,39 @@ class TimelineInner extends Component {
...
+      pageNo: this.props.totalPages
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimeLine.js]:
```suggestion
```

File: client/src/components/profile/components/TimeLine.js [hunk 10]
```diff
@@ -81,9 +108,39 @@ class TimelineInner extends Component {
...
+    });
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/TimeLine.js]:
```suggestion
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Thanks @MathewtheCoder that's what I was looking for.

One last nitpick:

**Comment by MathewtheCoder** [comment]
@ojeytonwilliams Changes you requested have been done.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1: 

Thanks for sticking with this @MathewtheCoder

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
@moT01 and @RandellDawson can you sign off on this, before I demo this to Quincy and get this landed.

**Comment by RandellDawson** [comment]
I will take a look at this one today.

File: client/src/components/profile/components/TimeLine.js [hunk 11]
```diff
@@ -1,13 +1,18 @@
-import React, { Component } from 'react';
+import React, { Component, useMemo } from 'react';
 import PropTypes from 'prop-types';
 import format from 'date-fns/format';
-import { find, reverse, sortBy } from 'lodash';
+import find from 'lodash/find';
+import reverse from 'lodash/reverse';
+import sortBy from 'lodash/sortBy';
```
Comment (moT01) [inline client/src/components/profile/components/TimeLine.js]: Wondering if there's a reason that there were separated? Looked cleaner how it was to me.

File: client/src/components/profile/components/TimelinePagination.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const TimelinePagination = props => {
+  const { pageNo, totalPages, firstPage, prevPage, nextPage, lastPage } = props;
+  return (
+    <nav
+      aria-label='Timeline Pagination Navigation'
```
Comment (moT01) [inline client/src/components/profile/components/TimelinePagination.js]:
```suggestion
      aria-label='Timeline Pagination'
```
See pics.

File: client/src/components/profile/components/TimelinePagination.js [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+      className='timeline-pagination'
```
Comment (moT01) [inline client/src/components/profile/components/TimelinePagination.js]:
```suggestion
```
I don't see this class being used, do we need it?

File: client/src/components/profile/components/TimelinePagination.js [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+      role='navigation'
+    >
+      <ul className='timeline-pagination_list'>
```
Comment (moT01) [inline client/src/components/profile/components/TimelinePagination.js]:
```suggestion
      <ul aria-hidden='true' className='timeline-pagination_list'>
```
see pics.

**Review by moT01 — COMMENTED** [review]
Functionally, this LGTM @MathewtheCoder. The buttons all work smooth and I would be fine adding it like this.

I've been looking at accessibility lately though, so I focused on that part of the code. I'm not an expert in this area, but I turned my mac's built in screen reader on and got this from the buttons...
Screen Shot 2019-11-10 at 9 21 30 AM

Adding my suggestions turns it into this...
Screen Shot 2019-11-10 at 9 22 39 AM

I think it's more understandable like this. I'm not sure how other screen readers will interpret these things though and I'm not sure how someone using one will expect it to be.

File: client/src/components/profile/components/TimeLine.js (hunk 11 above)
Comment (MathewtheCoder) [reply client/src/components/profile/components/TimeLine.js]: @moT01 I saw in some components that only needed functions were being imported from lodash and though the change would be brought to all components. Thats why i did like this.

**Comment by MathewtheCoder** [comment]
@moT01 Made the changes

File: client/src/components/profile/components/TimeLine.js (hunk 11 above)
Comment (moT01) [reply client/src/components/profile/components/TimeLine.js]: I thought both versions were doing the same thing.

**Review by moT01 — APPROVED** [review]
LGTM! 🎉

**Comment by MathewtheCoder** [comment]
@raisedadead Any updates on this PR ?

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
@raisedadead There are two earlier PRs (see below) which make changes to the `client/src/components/profile/components/TimeLine.js ` file. 

#35699 - this PR already has merge conflicts and is actually in a stale state at this point due to no activity from the contributor.  The changes in the current PR will the same section of code of the file above on this PR.
#36747 - is still in draft mode, so this should not be an issue

I am going to go ahead and merge, since the first earlier PR already has merge conflicts and is stale and the second PR should not be affected (merge conflicts).

**Comment by RandellDawson** [comment]
Thank you for your contribution to the page! 👍
We're happy to accept these changes, and look forward to future contributions. 🎉
